docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+
+Here's one example using the function `binaryPlus`:
+
+```purs
+binaryPlus :: Int -> Int -> Int
+binaryPlus a b = a + b
+
+usage = 9 + binaryPlus 1 2
+```
+
+There's a few ways we could inline the expression, `binaryPlus 1 2`. While `purs-backend-es` will performance its own analysis to determine whether and when to inline it, it doesn't know as much as the developer does.
+
+A developer can use **inline directives** to tell `purs-backend-es` whether to inline a function's body to its usage site and when (i.e. how many args need to be passed to the function before the inlining occurs):
+
+- never inline it at all (i.e. the directive is `never`)
+    - `9 + binaryPlus 1 2`
+- always inline it with no consideration for the number of arguments applied to it (i.e. the directive is `always`)
+    - `9 + (\a b -> a + b) 1 2`
+- inline it only after at least one argument has been applied to it (i.e. the directive is `arity=1`)
+    - `9 + (\  b -> 1 + b)   2`
+- inline it only after at least two arguments have been applied to it (i.e. the directive is `arity=2`)
+    - `9 + (        1 + 2)`
+
+**To summarize, inline directives provide developers with a stronger guarantee about whether and when a function is inlined than just relying upon the analysis done by `purs-backend-es`.**
+
+#### Inlining Duplicates Code
+
+The below example highlights the risk of inlining everything without thought: pointless code duplication.
+
+```purs
+hasExpensiveLetBinding a b = do
+  let
+    someValue = expensiveComputation
+  1 + someValue * b + a
+
+usage1 = hasExpensiveLetBinding 1 2
+usage2 = hasExpensiveLetBinding 3 4
+```
+
+If we inline `hasExpensiveLetBinding` immediately, the `expensiveComputation` will be computed three times rather than once. In other words, it would be the same as writing the following in source code:
+
+```purs
+hasExpensiveLetBinding a b = do
+  let
+    someValue = expensiveComputation
+  1 + someValue * b + a
+
+usage1 = (\a b -> do
+    let
+      someValue = expensiveComputation
+    1 + someValue * b + a
+  ) 1 2
+usage2 = (\a b -> do
+    let
+      someValue = expensiveComputation
+    1 + someValue * b + a
+  ) 3 4
+```
+
+**To summarize, inlining for the sake of inlining is a great way to unnecessarily increase your program's bundle size**.
+
+#### Rewrite Rules
+
+A **rewrite rule** will replace one expression, usually involving literal values, with an equivalent expression. For example, an expression like `1 + 2` will always produce the value `3`. So, we can replace that expression with `3`. The rewrite rule here could be defined as, "Whenever the `+` function is called on two literal `Int` values, replace that expression with the sum of the two `Int` values."
+
+```purs
+-- Before
+foo = 1 + 2
+  --  ^^^^^ Rewrite rule: "I can optimize that! `1 + 2` is `3`"
```

Comment:
Ie, the optimizer knows how to evaluate primitive addition, and it sees that `1 + 2` can be reduced to `3`.